README.md
logfilestream
=========

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][cov-image]][cov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/logfilestream.svg?style=flat-square
[npm-url]: https://npmjs.org/package/logfilestream
[travis-image]: https://img.shields.io/travis/node-modules/logfilestream.svg?style=flat-square
[travis-url]: https://travis-ci.org/node-modules/logfilestream
[cov-image]: http://codecov.io/github/node-modules/logfilestream/coverage.svg?branch=master
[cov-url]: http://codecov.io/github/node-modules/logfilestream?branch=master
[david-image]: https://img.shields.io/david/node-modules/logfilestream.svg?style=flat-square
[david-url]: https://david-dm.org/node-modules/logfilestream
[snyk-image]: https://snyk.io/test/npm/logfilestream/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/logfilestream
[download-image]: https://img.shields.io/npm/dm/logfilestream.svg?style=flat-square
[download-url]: https://npmjs.org/package/logfilestream

Log file stream, including auto rolling feature, support multiprocess `append` write at the same time.

## Install

```sh
$ npm i logfilestream --save
```

## Usage

```js
var writestream = logfilestream({
  logdir: '/tmp/logs',
  nameformat: '[info.]YYYY-MM-DD[.log]',
});

writestream.write('hello');
writestream.write(' world\n');
writestream.end();
```

## Options

```js
/**
 * Log stream, auto cut the log file.
 *
 * log file name is concat with `prename + format + ext`.
 *
 * @param  {Object} options
 *  - {String} logdir, this dir must exists.
 *  - {String} nameformat, default is '[info.]YYYY-MM-DD[.log]',
 *    @see moment().format(): http://momentjs.com/docs/#/displaying/format/
 *    Also support '{pid}' for process pid.
 *  - {String} [encoding], default is utf-8, other encoding will encode by iconv-lite
 *  - {Number} [duration], default is one houre(24 * 3600000 ms), must >= 60s.
 *  - {String} [mode], default is '0666'.
 *  - {Number} [buffer] buffer duration, default is 1000ms
 *  - {Boolean} [mkdir] try to mkdir in each cut, make sure dir exist.
 *    useful when your nameformat like 'YYYY/MM/DD/[info.log]'.
 * return {LogStream}
 */
```

## License

(The MIT License)

Copyright(c) 2012 - 2014 fengmk2 <fengmk2@gmail.com>.
Copyright(c) 2015 - present node-modules and other contributors.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

1.1.0 / 2017-03-07
==================

  * feat: use cork/uncork to buffer data (#7)

1.0.2 / 2016-06-26
==================

  * fix: upgrade moment to 2.11.2 (#6)

1.0.1 / 2014-09-18
==================

 * since moment().add(period, number) is deprecated from moment 2.8.x. update related code with moment().add(number, period), to avoid warning. (@threeday0905)

1.0.0 / 2014-09-15
==================

 * stream.end() should also clear the flush interval. (@threeday0905)

0.1.3 / 2014-07-08
==================

  * expose flush method

0.1.2 / 2014-07-08
==================

 * support encoding
 * ignore coveralls

0.1.1 / 2014-07-07
==================

 * add options.mkdir
 * use mkdirp to make sure dir exist
 * use autod
 * fix typo
 * add npm image

0.1.0 / 2013-11-08
==================

  * buffer flush log
  * support coveralls
  * use jscover
  * update logo url
  * add logo


test/logstream.js

var should = require('should');
var logstream = require('../');
var moment = require('moment');
var path = require('path');
var fs = require('fs');
var mkdirp = require('mkdirp');
var exec = require('child_process').exec;
var existsSync = fs.existsSync || path.existsSync;

var logdir = path.join(__dirname, 'logstream-test');

describe('lib/logstream.js', function () {
  before(function (done) {
    exec(['rm -rf ' + logdir], function () {
      mkdirp(logdir, done);
    });
  });
  
  describe('createStream()', function () {
    it('should init and call this first cut() to create a writestream', function () {
      var stream = logstream({logdir: logdir, prename: 'info.'});
      stream.duration.should.equal(3600000);
      stream.logdir.should.equal(logdir);
      stream.format.should.equal('YYYY-MM-DD');
      stream.ext.should.equal('.log');
      stream.streamMode.should.equal('0666');
      stream.stream.path.should.equal(path.join(logdir,
        'info.' + moment().format(stream.format) + '.log'));
      stream.should.have.property('_reopening', true);
      stream.end();
      stream.should.have.property('stream', null);
    });
  });

  describe('firstDuration()', function () {
    it('should default one hour when !duration', function () {
      var stream = logstream({logdir: logdir, prename: 'info.'});
      stream.firstDuration().should.below(3600000);
      stream.end();

      [0, null, undefined].forEach(function (d) {
        var stream = logstream({duration: d, logdir: logdir, prename: 'info.'});
        stream.firstDuration().should.below(3600000);
        stream.end();
      });
    });

    it('should return <= duration', function () {
      [1, 10, 100, 999, 59999, 60000].forEach(function (i) {
        var stream = logstream({ duration: i, logdir: logdir });
        stream.firstDuration().should.equal(60000);
        stream.end();
      });
      [60000 * 5, 60000 * 10, 60000 * 20, 
        60000 * 30, 60000 * 59, 
        3600000 * 1, 3600000 * 2, 3600000 * 3, 
        3600000 * 5, 3600000 * 12, 3600000 * 24, 
        3600000 * 24 * 3, 3600000 * 24 * 5, 3600000 * 24 * 7
      ].forEach(function (j) {
        var stream = logstream({ duration: j, logdir: logdir });
        stream.firstDuration().should.below(j);
        stream.end();
      });
    });
  });

  describe('startTimer()', function () {
    it('_timer should be change', function () {
      var stream = logstream({logdir: logdir});
      var oldtimer = stream._timer;
      oldtimer.should.be.ok;
      stream.startTimer();
      stream._timer.should.not.equal(oldtimer);
      stream.end();
    });
  });

  describe('cut()', function () {
    var old = logstream.LogStream.prototype.firstDuration;
    before(function () {
      logstream.LogStream.prototype.firstDuration = function () {
        return 1000;
      };
    });
    after(function () {
      logstream.LogStream.prototype.firstDuration = old;
    });

    it('should be cut when timeout', function (done) {
      var stream = logstream({
        duration: 1000,
        format: 'YYYY-MM-DD hh:mm:ss',
        prename: 'debug.',
        logdir: logdir
      });
      var lastStream = stream.stream;
      var line1 = 'now is ' + new Date() + '|\n';
      stream.write(line1);
      stream.write(line1);
      setTimeout(function () {
        fs.readFileSync(lastStream.path, 'utf8').should.equal(line1 + line1);
        existsSync(stream.stream.path).should.be.true;
        stream.stream.path.should.not.equal(lastStream.path);
        var filepath = stream.stream.path;
        fs.readFileSync(stream.stream.path, 'utf8').should.equal('');
        stream.write(line1);
        stream.on('close', function () {
          fs.readFileSync(filepath, 'utf8').should.equal(line1 + line1 + line1);
          done();
        });
        stream.write(line1);
        stream.write(line1);
        stream.end();
        stream.should.have.property('_timer', null);
        stream.should.have.property('stream', null);
      }, 1010);
    });
  });
});

test/logstream.test.js
'use strict';

var should = require('should');
var moment = require('moment');
var path = require('path');
var fs = require('fs');
var mkdirp = require('mkdirp');
var exec = require('child_process').exec;
var existsSync = fs.existsSync || path.existsSync;
var logstream = require('../');

var logdir = path.join(__dirname, 'logstream-test');

describe('logstream.test.js', function () {
  before(function (done) {
    exec(['rm -rf ' + logdir], function () {
      mkdirp(logdir, done);
    });
  });

  describe('createStream()', function () {
    it('should init and call this first cut() to create a writestream', function () {
      var stream = logstream({logdir: logdir});
      stream.duration.should.equal(3600000);
      stream.logdir.should.equal(logdir);
      stream.nameformat.should.equal('[info.]YYYY-MM-DD[.log]');
      stream.streamMode.should.equal('0666');
      stream.stream.path.should.equal(path.join(logdir,
        'info.' + moment().format('YYYY-MM-DD') + '.log'));
      stream.should.have.property('_reopening', true);
      stream.end();
      stream.should.have.property('stream', null);
    });

    it('should init format YYYY/MM/DD/[info.log] ok', function () {
      var stream = logstream({logdir: logdir, nameformat: 'YYYY/MM/DD/[info.log]', mkdir: true});
      stream.duration.should.equal(3600000);
      stream.logdir.should.equal(logdir);
      stream.nameformat.should.equal('YYYY/MM/DD/[info.log]');
      stream.streamMode.should.equal('0666');
      stream.stream.path.should.equal(path.join(logdir,
        moment().format('YYYY/MM/DD/') + 'info.log'));
      stream.should.have.property('_reopening', true);
      stream.end();
      stream.should.have.property('stream', null);
    })
  });

  describe('firstDuration()', function () {
    it('should default one hour when !duration', function () {
      var stream = logstream({logdir: logdir, prename: 'info.'});
      stream.firstDuration().should.below(3600000);
      stream.end();

      [0, null, undefined].forEach(function (d) {
        var stream = logstream({duration: d, logdir: logdir, prename: 'info.'});
        stream.firstDuration().should.below(3600000);
        stream.end();
      });
    });

    it('should return <= duration', function () {
      [1, 10, 100, 999, 59999, 60000].forEach(function (i) {
        var stream = logstream({ duration: i, logdir: logdir });
        stream.firstDuration().should.equal(60000);
        stream.end();
      });
      [60000 * 5, 60000 * 10, 60000 * 20,
        60000 * 30, 60000 * 59,
        3600000 * 1, 3600000 * 2, 3600000 * 3,
        3600000 * 5, 3600000 * 12, 3600000 * 24,
        3600000 * 24 * 3, 3600000 * 24 * 5, 3600000 * 24 * 7
      ].forEach(function (j) {
        var stream = logstream({ duration: j, logdir: logdir });
        stream.firstDuration().should.below(j);
        stream.end();
      });
    });
  });

  describe('startTimer()', function () {
    it('_timer should be change', function () {
      var stream = logstream({logdir: logdir});
      var oldtimer = stream._timer;
      oldtimer.should.be.ok;
      stream.startTimer();
      stream._timer.should.not.equal(oldtimer);
      stream.end();
    });
  });

  describe('cut()', function () {
    var old = logstream.LogStream.prototype.firstDuration;
    before(function () {
      logstream.LogStream.prototype.firstDuration = function () {
        return 1000;
      };
    });
    after(function () {
      logstream.LogStream.prototype.firstDuration = old;
    });

    it('should be cut when timeout', function (done) {
      var stream = logstream({
        duration: 1000,
        nameformat: 'YYYY-MM-DDhh:mm:ss[.log]',
        logdir: logdir
      });
      var line1 = 'now is ' + new Date() + '|\n';
      stream.write(line1);
      stream.write(line1);
      var lastStream = stream.stream;
      setTimeout(function () {
        fs.readFileSync(lastStream.path, 'utf8').should.equal(line1 + line1);
        existsSync(stream.stream.path).should.be.true;
        stream.stream.path.should.not.equal(lastStream.path);
        var filepath = stream.stream.path;
        fs.readFileSync(stream.stream.path, 'utf8').should.equal('');
        stream.write(line1);
        stream.on('close', function () {
          fs.readFileSync(filepath, 'utf8').should.equal(line1 + line1 + line1);
          done();
        });
        stream.write(line1);
        stream.write(line1);
        stream.end();
        stream.should.have.property('_timer', null);
        stream.should.have.property('stream', null);
      }, 1500);
    });
  });

  describe('end()', function() {
    it('will clear timer and flush timer', function() {
      var stream = logstream({logdir: logdir});

      stream.should.have.property('_timer').be.a.Object;
      stream.should.have.property('_flushTimer').be.a.Object;

      stream.end();

      stream.should.have.property('_timer', null);
      stream.should.have.property('_flushTimer', null);
    });

  });
});


index.js
'use strict';

var fs = require('fs');
var path = require('path');
var EventEmitter = require('events').EventEmitter;
var util = require('util');
var moment = require('moment');
var mkdirp = require('mkdirp');
var iconv = require('iconv-lite');

var ONE_MINUTE = 60000;
var ONE_HOUR = 60 * ONE_MINUTE;
var ONE_DAY = 24 * ONE_HOUR;

/**
 * Default log buffer duration.
 */
var defaultBufferDuration = 1000;

/**
 * Log stream, auto cut the log file.
 *
 * log file name is concat with `prename + format + ext`.
 *
 * @param  {Object} options
 *  - {String} logdir, this dir must exists.
 *  - {String} nameformat, default is '[info.]YYYY-MM-DD[.log]',
 *    @see moment().format(): http://momentjs.com/docs/#/displaying/format/
 *    Also support '{pid}' for process pid.
 *  - {String} [encoding], default is utf-8, other encoding will encode by iconv-lite
 *  - {Number} [duration], default is one houre(24 * 3600000 ms), must >= 60s.
 *  - {String} [mode], default is '0666'.
 *  - {Number} [buffer] buffer duration, default is 1000ms
 *  - {Boolean} [mkdir] try to mkdir in each cut, make sure dir exist.
 *    useful when your nameformat like 'YYYY/MM/DD/[info.log]'.
 * return {LogStream}
 */
module.exports = function createStream(options) {
  return new LogStream(options);
};

function LogStream(options) {
  if (!(this instanceof LogStream)) {
    return new LogStream(options);
  }
  options = options || {};
  this.logdir = options.logdir;
  this.nameformat = options.nameformat || '[info.]YYYY-MM-DD[.log]';
  this.nameformat = this.nameformat.replace('{pid}', process.pid);
  this.duration = options.duration || ONE_HOUR;
  // must >= one minute
  if (this.duration < 60000) {
    this.duration = 60000;
  }
  this.encoding = (options.encoding || 'utf-8').toLowerCase();
  if (this.encoding === 'utf8') {
    this.encoding = 'utf-8';
  }

  this.streamMode = options.mode || '0666';
  this.mkdir = options.mkdir;
  this.cut();
  this.startTimer(this.firstDuration());
  this._flushInterval = options.buffer || defaultBufferDuration;
  this._flushTimer = setInterval(this._flush.bind(this), this._flushInterval);
}

util.inherits(LogStream, EventEmitter);

LogStream.prototype.firstDuration = function () {
  var firstDuration = this.duration;
  if (this.duration > ONE_MINUTE) {
    var now = moment();
    if (this.duration < ONE_HOUR) { // in minute
      firstDuration = now.clone().add(this.duration, 'ms').startOf('minute').diff(now);
    } else if (this.duration < ONE_DAY) { // in hour
      firstDuration = now.clone().add(this.duration, 'ms').startOf('hour').diff(now);
    } else { // in day
      firstDuration = now.clone().add(this.duration, 'ms').startOf('day').diff(now);
    }
  }
  return firstDuration;
};

LogStream.prototype.startTimer = function (duration) {
  this._timer = setTimeout(function (self) {
    self.cut();
    self.startTimer(self.duration);
  }, duration || this.duration, this);
};

LogStream.prototype.cut = function () {
  if (this.stream) {
    this._flush();
    this.stream.end();
    this.stream.destroySoon();
    this.stream = null;
  }
  var name = moment().format(this.nameformat);
  var logpath = path.join(this.logdir, name);

  // make sure dir exist
  if (this.mkdir) {
    try {
      mkdirp.sync(path.dirname(logpath));
    } catch (err) {
      // ignore
    }
  }

  this._reopening = true;
  this.stream = fs.createWriteStream(logpath, {flags: 'a', mode: this.streamMode});
  this.stream.cork();
  this.stream
    .on("error", this.emit.bind(this, "error"))
    .on("pipe", this.emit.bind(this, "pipe"))
    .on("drain", this.emit.bind(this, "drain"))
    .on("open", function () {
      this._reopening = false;
    }.bind(this))
    .on("close", function () {
      if (!this._reopening) {
        this.emit("close");
      }
    }.bind(this));
};

LogStream.prototype.write = function (string) {
  this.stream.write(this._encode(string));
};

LogStream.prototype.flush =
LogStream.prototype._flush = function () {
  this.stream.uncork();
  this.stream.cork();
};

LogStream.prototype._encode = function (string) {
  if (this.encoding === 'utf-8') {
    return new Buffer(string);
  }
  return iconv.encode(string, this.encoding);
};

LogStream.prototype.end = function () {
  if (this._timer) {
    clearTimeout(this._timer);
    this._timer = null;
  }

  if (this._flushTimer) {
    clearInterval(this._flushTimer);
    this._flushTimer = null;
  }

  if (this.stream) {
    this._flush();
    this.stream.end();
    this.stream.destroySoon();
    this.stream = null;
  }
};

LogStream.prototype.close = LogStream.prototype.end;

module.exports.LogStream = LogStream;


lib/logstream.js
/**!
 * logstream - lib/logstream.js
 *
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var fs = require('fs');
var path = require('path');
var EventEmitter = require('events').EventEmitter;
var util = require('util');
var moment = require('moment');
var mkdirp = require('mkdirp');
var iconv = require('iconv-lite');

var ONE_MINUTE = 60000;
var ONE_HOUR = 60 * ONE_MINUTE;
var ONE_DAY = 24 * ONE_HOUR;

/*!
 * Default log buffer duration.
 */
var defaultBufferDuration = 1000;

/**
 * Log stream, auto cut the log file.
 *
 * log file name is concat with `prename + format + ext`.
 *
 * @param  {Object} options
 *  - {String} logdir, this dir must exists.
 *  - {String} nameformat, default is '[info.]YYYY-MM-DD[.log]',
 *    @see moment().format(): http://momentjs.com/docs/#/displaying/format/
 *    Also support '{pid}' for process pid.
 *  - {String} [encoding], default is utf-8, other encoding will encode by iconv-lite
 *  - {Number} [duration], default is one houre(24 * 3600000 ms), must >= 60s.
 *  - {String} [mode], default is '0666'.
 *  - {Number} [buffer] buffer duration, default is 1000ms
 *  - {Boolean} [mkdir] try to mkdir in each cut, make sure dir exist.
 *    useful when your nameformat like 'YYYY/MM/DD/[info.log]'.
 * return {LogStream}
 */
module.exports = function createStream(options) {
  return new LogStream(options);
};

function LogStream(options) {
  if (!(this instanceof LogStream)) {
    return new LogStream(options);
  }
  options = options || {};
  this.logdir = options.logdir;
  this.nameformat = options.nameformat || '[info.]YYYY-MM-DD[.log]';
  this.nameformat = this.nameformat.replace('{pid}', process.pid);
  this.duration = options.duration || ONE_HOUR;
  // must >= one minute
  if (this.duration < 60000) {
    this.duration = 60000;
  }
  this.encoding = (options.encoding || 'utf-8').toLowerCase();
  if (this.encoding === 'utf8') {
    this.encoding = 'utf-8';
  }

  this.streamMode = options.mode || '0666';
  this.mkdir = options.mkdir;
  this.cut();
  this.startTimer(this.firstDuration());
  this._buf = [];
  this._flushInterval = options.buffer || defaultBufferDuration;
  this._flushTimer = setInterval(this._flush.bind(this), this._flushInterval);
}

util.inherits(LogStream, EventEmitter);

LogStream.prototype.firstDuration = function () {
  var firstDuration = this.duration;
  if (this.duration > ONE_MINUTE) {
    var now = moment();
    if (this.duration < ONE_HOUR) { // in minute
      firstDuration = now.clone().add('ms', this.duration).startOf('minute').diff(now);
    } else if (this.duration < ONE_DAY) { // in hour
      firstDuration = now.clone().add('ms', this.duration).startOf('hour').diff(now);
    } else { // in day
      firstDuration = now.clone().add('ms', this.duration).startOf('day').diff(now);
    }
  }
  return firstDuration;
};

LogStream.prototype.startTimer = function (duration) {
  this._timer = setTimeout(function (self) {
    self.cut();
    self.startTimer(self.duration);
  }, duration || this.duration, this);
};

LogStream.prototype.cut = function () {
  if (this.stream) {
    this._flush();
    this.stream.end();
    this.stream.destroySoon();
    this.stream = null;
  }
  var name = moment().format(this.nameformat);
  var logpath = path.join(this.logdir, name);

  // make sure dir exist
  if (this.mkdir) {
    try {
      mkdirp.sync(path.dirname(logpath));
    } catch (err) {
      // ignore
    }
  }

  this._reopening = true;
  this.stream = fs.createWriteStream(logpath, {flags: 'a', mode: this.streamMode});
  this.stream
    .on("error", this.emit.bind(this, "error"))
    .on("pipe", this.emit.bind(this, "pipe"))
    .on("drain", this.emit.bind(this, "drain"))
    .on("open", function () {
      this._reopening = false;
    }.bind(this))
    .on("close", function () {
      if (!this._reopening) {
        this.emit("close");
      }
    }.bind(this));
};

LogStream.prototype.write = function (string) {
  this._buf.push(string);
};

LogStream.prototype.flush =
LogStream.prototype._flush = function () {
  if (this._buf.length) {
    var buf = this._encode(this._buf.join(''));
    this.stream.write(buf);
    this._buf.length = 0;
  }
};

LogStream.prototype._encode = function (string) {
  if (this.encoding === 'utf-8') {
    return new Buffer(string);
  }
  return iconv.encode(string, this.encoding);
};

LogStream.prototype.end = function () {
  if (this._timer) {
    clearTimeout(this._timer);
    this._timer = null;
  }

  if (this._flushTimer) {
    clearInterval(this._flushTimer);
    this._flushTimer = null;
  }

  if (this.stream) {
    this._flush();
    this.stream.end();
    this.stream.destroySoon();
    this.stream = null;
  }
};

LogStream.prototype.close = LogStream.prototype.end;

module.exports.LogStream = LogStream;
